src/material-experimental/mdc-form-field/form-field.ts

```diff
@@ -0,0 +1,485 @@
+    // form-field because we support abstract form field controls which are not necessarily
+    // of type input, nor do we have a reference to a native form-field control element. Instead
+    // we handle the focus by checking if the abstract form-field control focused state changes.
+    if (this._control.focused && !this._isFocused) {
+      this._isFocused = true;
+      this._foundation.activateFocus();
+    } else if (!this._control.focused && (this._isFocused || this._isFocused === null)) {
+      this._isFocused = false;
+      this._foundation.deactivateFocus();
+    }
+  }
+
+  _rerenderOutlineNotch() {
+    if (this._floatingLabel && this._hasOutline()) {
+      this._foundation.notchOutline(this._shouldLabelFloat());
+    }
+  }
+
+  /** Whether the floating label should always float or not. */
+  _shouldAlwaysFloat() {
+    return this.floatLabel === 'always';
+  }
+
+  _hasOutline() {
+    return this.appearance === 'outline';
+  }
+
+  _hasFloatingLabel() {
+    return !!this._labelChildNonStatic || !!this._labelChildStatic;
+  }
+
+  _shouldLabelFloat() {
+    return this._control.shouldLabelFloat || this._shouldAlwaysFloat();
+  }
+
+  _isStandardAppearance() {
+    return this.appearance === 'standard' || this.appearance === 'legacy';
+  }
+
+  /** Determines whether a class from the NgControl should be forwarded to the host element. */
+  _shouldForward(prop: keyof NgControl): boolean {
+    const ngControl = this._control ? this._control.ngControl : null;
+    return ngControl && ngControl[prop];
+  }
+
+  /** Determines whether to display hints or errors. */
+  _getDisplayedMessages(): 'error' | 'hint' {
+    return (this._errorChildren && this._errorChildren.length > 0 &&
+      this._control.errorState) ? 'error' : 'hint';
+  }
+
+  /** Does any extra processing that is required when handling the hints. */
+  private _processHints() {
+    this._validateHints();
+    this._syncDescribedByIds();
+  }
+
+  /**
+   * Ensure that there is a maximum of one of each `<mat-hint>` alignment specified, with the
+   * "hintLabel" attribute being considered as `align="start"`.
+   */
+  private _validateHints() {
+    if (this._hintChildren) {
+      let startHint: MatHint;
+      let endHint: MatHint;
+      this._hintChildren.forEach((hint: MatHint) => {
+        if (hint.align === 'start') {
+          if (startHint || this.hintLabel) {
+            throw getMatFormFieldDuplicatedHintError('start');
+          }
+          startHint = hint;
+        } else if (hint.align === 'end') {
+          if (endHint) {
+            throw getMatFormFieldDuplicatedHintError('end');
+          }
+          endHint = hint;
+        }
+      });
+    }
+  }
+
+  /**
+   * Sets the list of element IDs that describe the child control. This allows the control to update
+   * its `aria-describedby` attribute accordingly.
+   */
+  private _syncDescribedByIds() {
```

Comment:
Would using `AriaDescriber` simplify this method at all?

Reply:
I don't think so. We want to use existing DOM elements and the actual custom form-field control is responsible for setting them. We just pass the hint/error id's down to the registered control.